snooty/parser.py

```diff
@@ -940,6 +944,21 @@ def validate_and_add_asset(
                 CannotOpenFile(Path(image_argument), err.strerror, line)
             )
 
+    def validate_relative_url(self, url_argument: str, line: int) -> None:
+        """Validate relative URL points to page within current docs site.
+        URLs can be of the form /foo, foo, /foo/"""
+        target_path = util.add_doc_target_ext(
+            url_argument, self.docpath, self.project_config.source_path
+        )
+
+        if not target_path.is_file():
+            err_message = (
+                f"{os.strerror(errno.ENOENT)} for relative path {url_argument}"
+            )
+            self.diagnostics.append(CannotOpenFile(target_path, err_message, line))
+        elif re.search(r"([\/])\1", url_argument) is not None:
```

Comment:
It's best to precompile regexes with `re.compile()` in the global scope: the python `re` module *does* maintain a cache, but when that cache reaches its maximum size (512 entries) the cache is wiped.

This normally isn't an issue, and shouldn't be in snooty, but in a larger application it means that dynamically constructed regexes can end up wiping out performance elsewhere unnecessarily.

Reply:
TIL! thanks for sharing your Python knowledge! 🧠